crates/dbsp/src/trace/spine_fueled.rs

```diff
@@ -231,6 +233,51 @@ where
         self.lower_key_bound = Some(bound.clone());
         self.map_batches_mut(|batch| batch.truncate_keys_below(&bound));
     }
+
+    fn sample_keys<RG>(&self, rng: &mut RG, sample_size: usize, sample: &mut Vec<Self::Key>)
+    where
+        RG: Rng,
+        Self::Time: PartialEq<()>,
+    {
+        let total_keys = self.key_count();
+
+        if sample_size == 0 || total_keys == 0 {
+            // Avoid division by zero.
+            return;
+        }
+
+        // Sample each batch, picking the number of keys proportional to
+        // batch size.
+        let mut intermediate = Vec::with_capacity(sample_size);
+
+        self.map_batches(|batch| {
+            batch.sample_keys(
+                rng,
+                batch.key_count() * sample_size / total_keys,
+                &mut intermediate,
+            );
+        });
+
+        // Drop duplicate keys and keys that appear with 0 weight, i.e.,
+        // get canceled out across multiple batches.
+        intermediate.sort_unstable();
+        intermediate.dedup();
```

Comment:
Should we call `consolidate()` here?

Reply:
`consolidate` is good for tuples.  This one's a simple array.